File: 프로그래머스/Level 2/피로도-review2-solve2.java
class Solution {
    public int solution(int k, int[][] dungeons) {
        int maxCompletedDungeonsNum = dfs(k, 0, dungeons, 
																					new boolean[dungeons.length]);
        return maxCompletedDungeonsNum;
    }
    
    private int dfs(int nowFatigue, int completedDungeons, 
										int[][] dungeons, boolean[] visited) {
        int maxCompletedDungeonsNum = completedDungeons;
        
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && nowFatigue >= dungeons[i][0]) {
                visited[i] = true;
                int result = dfs(nowFatigue - dungeons[i][1], 
																	completedDungeons + 1, dungeons, visited);
                maxCompletedDungeonsNum = 
																		Math.max(maxCompletedDungeonsNum, result);
                visited[i] = false;
            }
        }
        
        return maxCompletedDungeonsNum;
    }
}

File: 프로그래머스/Level 2/피로도-review2-solve2.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    
    def dfs(now_fatigue, completed_dungeons):
        nonlocal max_completed_dungeons_num
        
        for i in range(len(dungeons)):
            if not visited[i] and now_fatigue >= dungeons[i][0]:
                visited[i] = True
                dfs(now_fatigue - dungeons[i][1], completed_dungeons + 1)
                visited[i] = False
                
        max_completed_dungeons_num = max(max_completed_dungeons_num, 
																					completed_dungeons) 
        
    max_completed_dungeons_num = 0 
    
    dfs(k, 0)
    
    return max_completed_dungeons_num